File: VoiceType/Services/HotwordCorrector.swift
import Foundation

/// 识别结果后处理热词纠正：对文本按热词字数开滑窗，
/// 窗口拼音与热词拼音编辑距离 ≤ ⌈拼音长度×20%⌉（至少 1）即替换。
/// 仅处理 CJK 字符窗口（拼音方案对英文无意义）。
struct HotwordCorrector {
    private let entries: [(chars: [Character], pinyin: String)]

    init(hotwords: [String]) {
        entries = hotwords
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty && $0.allSatisfy(Self.isCJK) }
            .map { (Array($0), Self.pinyin(of: $0)) }
            .filter { !$0.1.isEmpty }
    }

    static func isCJK(_ c: Character) -> Bool {
        guard let scalar = c.unicodeScalars.first else { return false }
        return (0x4E00...0x9FFF).contains(scalar.value)
    }

    static func pinyin(of text: String) -> String {
        let mutable = NSMutableString(string: text)
        CFStringTransform(mutable, nil, kCFStringTransformMandarinLatin, false)
        CFStringTransform(mutable, nil, kCFStringTransformStripDiacritics, false)
        return (mutable as String).lowercased().replacingOccurrences(of: " ", with: "")
    }

    func correct(_ text: String) -> String {
        guard !entries.isEmpty else { return text }
        var chars = Array(text)
        for (target, targetPinyin) in entries {
            let n = target.count
            guard chars.count >= n else { continue }
            let maxDistance = max(1, Int((Double(targetPinyin.count) * 0.2).rounded(.up)))
            var i = 0
            while i + n <= chars.count {
                let window = Array(chars[i..<(i + n)])
                if window == target {
                    i += n
                    continue
                }
                guard window.allSatisfy(Self.isCJK) else {
                    i += 1
                    continue
                }
                let windowPinyin = Self.pinyin(of: String(window))
                if Self.levenshtein(targetPinyin, windowPinyin) <= maxDistance {
                    chars.replaceSubrange(i..<(i + n), with: target)
                    i += n
                } else {
                    i += 1
                }
            }
        }
        return String(chars)
    }

    static func levenshtein(_ a: String, _ b: String) -> Int {
        let x = Array(a.utf8), y = Array(b.utf8)
        if x.isEmpty { return y.count }
        if y.isEmpty { return x.count }
        var prev = Array(0...y.count)
        var curr = [Int](repeating: 0, count: y.count + 1)
        for i in 1...x.count {
            curr[0] = i
            for j in 1...y.count {
                let cost = x[i - 1] == y[j - 1] ? 0 : 1
                curr[j] = min(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + cost)
            }
            swap(&prev, &curr)
        }
        return prev[y.count]
    }
}

File: windows/src/VoiceType/Services/HotwordCorrector.cs
namespace VoiceType.Services;

/// <summary>
/// 汉字 → 无声调小写拼音（对应 macOS CFStringTransform MandarinLatin + StripDiacritics）。
/// 实现用 TinyPinyin.Net；抽象成接口便于替换与单测注入。
/// </summary>
public interface IPinyinProvider
{
    /// <summary>非汉字字符原样保留（与 CFStringTransform 行为一致，随后统一小写、去空白）。</summary>
    string Pinyin(string text);
}

public sealed class TinyPinyinProvider : IPinyinProvider
{
    public string Pinyin(string text) =>
        TinyPinyin.PinyinHelper.GetPinyin(text, "").ToLowerInvariant().Replace(" ", "");
}

/// <summary>
/// 识别结果后处理热词纠正：对文本按热词字数开滑窗，
/// 窗口拼音与热词拼音编辑距离 ≤ ⌈拼音长度×20%⌉（至少 1）即替换。
/// 仅处理 CJK 字符窗口（拼音方案对英文无意义）。
/// </summary>
public sealed class HotwordCorrector
{
    private readonly List<(char[] Chars, string Pinyin)> _entries;
    private readonly IPinyinProvider _pinyin;

    public HotwordCorrector(IEnumerable<string> hotwords)
        : this(hotwords, new TinyPinyinProvider()) { }

    public HotwordCorrector(IEnumerable<string> hotwords, IPinyinProvider pinyin)
    {
        _pinyin = pinyin;
        _entries = hotwords
            .Select(w => w.Trim())
            .Where(w => w.Length > 0 && w.All(IsCjk))
            .Select(w => (w.ToCharArray(), pinyin.Pinyin(w)))
            .Where(e => e.Item2.Length > 0)
            .ToList();
    }

    public static bool IsCjk(char c) => c is >= '\u4E00' and <= '\u9FFF';

    public string Correct(string text)
    {
        if (_entries.Count == 0)
            return text;
        char[] chars = text.ToCharArray();
        foreach ((char[] target, string targetPinyin) in _entries)
        {
            int n = target.Length;
            if (chars.Length < n)
                continue;
            int maxDistance = Math.Max(1, (int)Math.Ceiling(targetPinyin.Length * 0.2));
            int i = 0;
            while (i + n <= chars.Length)
            {
                var window = new char[n];
                Array.Copy(chars, i, window, 0, n);
                if (window.SequenceEqual(target))
                {
                    i += n;
                    continue;
                }
                if (!Array.TrueForAll(window, IsCjk))
                {
                    i += 1;
                    continue;
                }
                string windowPinyin = PinyinOf(window);
                if (Levenshtein(targetPinyin, windowPinyin) <= maxDistance)
                {
                    Array.Copy(target, 0, chars, i, n);
                    i += n;
                }
                else
                {
                    i += 1;
                }
            }
        }
        return new string(chars);
    }

    private string PinyinOf(char[] window) => _pinyin.Pinyin(new string(window));

    public static int Levenshtein(ReadOnlySpan<char> a, ReadOnlySpan<char> b)
    {
        if (a.IsEmpty) return b.Length;
        if (b.IsEmpty) return a.Length;
        int[] prev = Enumerable.Range(0, b.Length + 1).ToArray();
        var curr = new int[b.Length + 1];
        for (int i = 1; i <= a.Length; i++)
        {
            curr[0] = i;
            for (int j = 1; j <= b.Length; j++)
            {
                int cost = a[i - 1] == b[j - 1] ? 0 : 1;
                curr[j] = Math.Min(Math.Min(prev[j] + 1, curr[j - 1] + 1), prev[j - 1] + cost);
            }
            (prev, curr) = (curr, prev);
        }
        return prev[b.Length];
    }
}
